### PR TITLE
atomics: use explicit types for inter-process atomics

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

Inter process atomics have different semantics from normal atomics -- even though the internal mechanism may not have much difference when compiler native atomic support is available. It become significant when the native support is missing and we have to fall-back to lock-based atomics, the difference may become significant, especially during init/finalize.

By explicitly differentiating intra-process atomics and inter-process atomics, it makes the code semantically easier to read and also it makes atomic implementations and detail solutions more targeted and simpler.

Note this PR is supposed to be only the last 7 or so commits. All other commits will get rebased aways at some point.

## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [ ] Remove xfail from the test suite when fixing a test
* [ ] Commits are self-contained and do not do two things at once
* [ ] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [ ] Passes whitespace checkers
* [ ] Passes warning tests
* [ ] Passes all tests
* [ ] Add comments such that someone without knowledge of the code could understand
* [ ] Add Devel Docs in the `doc/` directory for any new code design

### CODE REVIEW

**Comment by pavanbalaji** [comment]
Yikes!  I have a half-done patch doing basically the same thing.  I guess I should have coordinated before working on it.  Oh, well.  Thanks for getting it done.

**Comment by hzhou** [comment]
> Yikes! I have a half-done patch doing basically the same thing. I guess I should have coordinated before working on it. Oh, well. Thanks for getting it done.

I am so glad that we are on the same page :)

**Comment by hzhou** [comment]
Tagging @shintaro-iwasaki . Only pay attention to the last 7 commits.

**Comment by pavanbalaji** [comment]
Hmm.  I missed the inter-process atomics part.  I saw the first few commits and thought it's the same thing that I was doing.  Can you point me to the discussion on that part?

**Comment by hzhou** [comment]
> Hmm. I missed the inter-process atomics part. I saw the first few commits and thought it's the same thing that I was doing. Can you point me to the discussion on that part?

It's an offline discussion with @shintaro-iwasaki . The jenkins tests for pgi compilers are failing because it falls back to lock-based atomics. Originally I thought was due to too slow. Then as I read about inter-process lock, I realize it must be due to MPL_atomc_ doesn't support inter-process lock. So I asked Shitaro to add the support. Then it gets to safety discussion, especially during finalization. How do we ensure other threads can still safely call `MPI_Initialized/Finalized` while the main thread destroys the inter-process lock? It seems impossible. The key issue is we mix intra-process atomics and and inter-process atomics. The intra-ones don't need initialization or finalization. Without differentiation, we can't tell which lock to use.

It also seems to be very good idea anyway to explicitly mark inter-process atomics, since they are used very differently from other atomics and worth drawing attentions even when the underlying mechanics are the same.

**Comment by hzhou** [comment]
Reference PR #4387

**Comment by pavanbalaji** [comment]
For lock-based atomics, pthreads already gives a flag for interprocess locks, which we already used to do in OPA.  If that got lost when we moved to MPL, that should be added back.  However, that's separate from the non-lock-based atomics.  It'll be good to have that discussion in a public channel (perhaps on slack) rather than in private.

**Comment by shintaro-iwasaki** [comment]
Introducing an interprocess lock.
Pros:
- Clearer semantics
As discussed in #4026, the current MPL/atomic and OPA do not really guarantee interprocess atomicity. We just "assume" compier-supported ones can use for inteprocess atomicity. In theory, they should be different, though.

    - Correctness
The lock-based option (which is prepared for a fallback) is the only implementation that explicitly supports interprocess atomicity (though it is slower than compiler-supported one). If needed, users can use this lock-based option only for interprocess atomicity while using (fast) compiler-supported one for intraprocess atomicity. This can maximize performance while maintaining correctness in a portable way.

    - Programming style
Intraprocess lock should not be used before shm initialization or after shm finalization. If we introduce an explicit intraprocess lock, developers can get more aware of this. 

- Performance optimization for emulation lock.
The current lock-based implementation (a fallback) always uses an interproecss lock for any atomic operations (i.e., synchronizing all processes in a node) even if that atomic operation only requires intraprocess atomicity (e.g., a completion counter of request objects). This differentiation can make this lock-based implementation more efficient; we can use an interprocess lock (i.e., a per-process global lock) for intraprocess atomicity .

Cons:
- It increases # of an atomic layer.
- It might increase some programming efforts.

**Comment by hzhou** [comment]
There is no issue with nonlock based atomics. Yes, we are essentially porting the inter process lock from opa. However,  the finalize issue seems valid that exist in openpa as well.

**Comment by pavanbalaji** [comment]
@shintaro-iwasaki Is the concern with always forcing inter-process locks that it'd lose performance?

**Comment by hzhou** [comment]
> It might increase some programming efforts.

It just require more precise datatype, so not much more programming effort, as the change in this PR shows. I really like the results that now we clealy can see whether an atomic variable is synchronizing interprocess or just between threads.

**Comment by shintaro-iwasaki** [comment]
Yes, the performance is lowered because the interprocess lock synchronizes all processes in the same node. Extremely speaking, if MPI is configured with a global lock, all MPI operations in the same node are serialized even though processes are different.

**Comment by pavanbalaji** [comment]
> Yes, the performance is lowered because the interprocess lock synchronizes all processes in the same node. ~Extremely speaking, if MPI is configured with the global lock thread-safety mode, all MPI operations in the same node are serialized even though processes are different.~ (EDIT) All the atomic operations in the same node are serialized even though processes are different.
> 
> EDIT: this is about the emulation lock. If compiler-supported one (e.g., GCC's `__atomic`) can be used and works fine for interprocess atomicity as well, it does not incur unintentional performance loss. All atomic operations are run in parallel if they are independent regardless of whether it requires interprocess atomicity or not.

Do we really care about this?  This is for lock-based atomics, which is really backup for us.  I don't like overcomplicating the code for this backup path.  And AFAICT, this only occurs with PGI when strict mode is enabled (because of one of the flags).  So it's not even the case that we have a compiler that shows this problem in a "real" case.

**Comment by hzhou** [comment]
Currently pgi compiler does not support atomics and will fallback to locks for all configuration.

--
Hui Zhou

________________________________

**Comment by shintaro-iwasaki** [comment]
That is true. I do not think we really care the performance of the emulation-lock mode, either.

In my opinion, this is basically for correctness. Although we have not encountered this case yet, if compiler-support one does not support interprocess atomicity, we can use efficiently use a correct one (i.e., emulation-lock) only for interprocess atomicity while we keep using compiler-supported one for intraprocess atomicity. If this point is not important, I would say the priority of introducing an explicit interprocess atomicity layer is low.

**Comment by pavanbalaji** [comment]
> Currently pgi compiler does not support atomics and will fallback to locks for all configuration.

The website seems to say that it supports C11: https://www.pgroup.com/resources/docs/17.10/x86/pgi-release-notes/index.htm

**Comment by shintaro-iwasaki** [comment]
I am not sure the really latest one, but when I tried last year, PGI did not support C11 atomics (as far as I checked).

**Comment by hzhou** [comment]
Partially supports C11 😀. That doc is from 2017. I tested 19.10, definitely no atomic support yet.

**Comment by pavanbalaji** [comment]
> That is true. I do not think we really care the performance of the emulation-lock mode, either.
> 
> In my opinion, this is basically for correctness. Although we have not encountered this case yet, if compiler-support one does not support interprocess atomicity, we can use efficiently use a correct one (i.e., emulation-lock) only for interprocess atomicity while we keep using compiler-supported one for intraprocess atomicity. If this point is not important, I would say the priority of introducing an explicit interprocess atomicity layer is low.

If the compiler supports atomics, we should use atomics everywhere (both intraprocess and interprocess).  If it doesn't support atomics, we should reevaluate how much we care about performance for that compiler.  If it's some old useless compiler that no one cares about, having worse performance is not a big deal (as long as it's functional).

I should also point out that PGI provides GCC-compiled library compatibility.  So, users can also build MPICH with GCC, even if they want to use PGI for their applications.

**Comment by shintaro-iwasaki** [comment]
Okay, so should we just create a (slow but) correct emulation lock as a fallback mainly for the current native PGI compilers? In any case, the current emulation lock wrapper is broken as @hzhou mentioned, so I need to fix it.

**Comment by hzhou** [comment]
So neither of you care about that more precise atomic type makes better code?😀

**Comment by pavanbalaji** [comment]
> So neither of you care about that more precise atomic type makes better code?😀

I just disagree with you on this.  I don't think it makes the code better.  I think makes it the code ugly.

**Comment by shintaro-iwasaki** [comment]
My standpoint is correctness; I am really willing to introduce this wrapper if a real compiler/OS combination breaks the following recommendation.(https://en.cppreference.com/w/cpp/atomic/atomic_is_lock_free)
```
The C++ standard recommends (but does not require) that lock-free
atomic operations are also address-free, that is, suitable for
communication between processes using shared memory.
```

We have not encounters such a case, though.

File: configure.ac [hunk 1]
```diff
@@ -1049,168 +1048,25 @@ else
...
-[with_zm_prefix=no])
-
-zmsrcdir=""
-AC_SUBST([zmsrcdir])
-zmlibdir=""
-AC_SUBST([zmlibdir])
-zmlib=""
-AC_SUBST([zmlib])
-
-if test "$enable_izem_queue" != "no" && test "$enable_izem_queue" != "none"; then
```
Comment (shawnccx) [inline configure.ac]: @pavanbalaji I remember you said only izem-atomic will be removed, `izem-queue` will still be available (when you were visiting Austin earlier we checked with you)

Here it seems like removing izem totally, including queue?

Comment (hzhou) [reply configure.ac]: The commits here are draft. This PR will be closed. We will opt for disable SHM when native atomics are unavailable

Comment (shawnccx) [reply configure.ac]: okay, thanks for clarification.

Comment (hzhou) [reply configure.ac]: izem_queue are safe for now :)

**Pull request closed without merging**